src/fields/extend_p.rs

```diff
@@ -84,47 +157,108 @@ where
+    {
+        *input_elem *= *kernel_elem;
+    }
 
-    // Precompute (-1)^i * (d choose i) * p(i).
-    let convolution_right_terms = nodes
+    // Transform the convolution result back.
+    let mut convolution_result = transformed_convolution_input;
+    FE::scaled_inverse_ntt_bit_reversed(&mut convolution_result);
+
+    for (convolution_elem, coeff) in convolution_result[nodes.len()..evaluations]
         .iter()
-        .enumerate()
-        .zip(context.binomial_coefficients.iter())
-        .map(|((i, p_i), binomial_coefficient)| {
-            let mut right = *binomial_coefficient * p_i;
-            // Multiply by (-1)^i. Note that it is safe to branch on i, since it doesn't depend on
-            // any secret.
-            if i & 1 != 0 {
-                right = -right;
-            }
-            right
+        .zip(context.after_convolution_coeffs.iter())
+    {
+        output.push(*convolution_elem * coeff);
+    }
+
+    output
+}
+
+/// Efficiently computes reciprocals of successive numbers.
+///
+/// The element at index zero is zero, then every other element is the reciprocal of its index.
+fn batched_inversion_sequence<FE: FieldElement>(length: usize) -> Vec<FE> {
+    if length == 0 {
+        return Vec::new();
+    }
+
+    let mut output = vec![FE::ZERO; length];
+
+    // First, compute the prefix products of successive numbers. The value of prefix_products[i]
+    // will be i!.
+    let mut prefix_products = Vec::with_capacity(length);
+    prefix_products.push(FE::ONE);
+    let mut product = FE::ONE;
+    let mut i = FE::ONE;
+    prefix_products.extend(
+        iter::repeat_with(|| {
+            product *= i;
+            i += FE::ONE;
+            product
         })
-        .collect::<Vec<FE>>();
+        .take(length - 1),
+    );
 
-    for k in nodes.len()..evaluations {
-        // Calculate (k - d) * (k choose d) from k from this iteration, and (k-1) choose d from the last iteration,
-        // using Remark A.3.
-        let k_minus_d_times_k_choose_d = FE::from_u128(k.try_into().unwrap()) * binomial_k_d;
-        // Update k choose d for k in this iteration, by dividing by (k - d).
-        binomial_k_d = k_minus_d_times_k_choose_d * context.reciprocals[k - d];
-        let sum = context.reciprocals[k - nodes.len() + 1..=k]
-            .iter()
-            .rev()
-            .zip(convolution_right_terms.iter())
-            .map(|(left, right)| *left * right)
-            .fold(FE::ZERO, |accumulator, value| accumulator + value);
-        let mut evaluation = k_minus_d_times_k_choose_d * sum;
-        // Multiply by (-1)^d. Note that it is safe to branch on d, since it is public knowledge.
-        if d & 1 != 0 {
-            evaluation = -evaluation;
-        }
-        output.push(evaluation);
+    // Take the multiplicative inverse of the last number. This will be ((length - 1)!)^-1.
+    let mut product_inverse = prefix_products.last().unwrap().mul_inv();
```

Comment:
```suggestion
    // Unwrap safety: prefix_products is guaranteed non-empty because length > 0
    let mut product_inverse = prefix_products.last().unwrap().mul_inv();
```